IHR2-2291 add documentation (#208)
dotwebstack/dotwebstack-framework: backend/rdf4j/README.MD

# 1. Backend-Rdf4j

## 1.1 Setup

The following snippet from an `application.yml` file shows the configuration properties for an Rdf4j backend:

```yml
dotwebstack:
  Rdf4j:
    shape:
      graph: https://github.com/dotwebstack/beer/shapes
      prefix: https://github.com/dotwebstack/beer/shapes#
    prefixes:
      rdf: http://www.w3.org/1999/02/22-rdf-syntax-ns#
      beer_def: https://github.com/dotwebstack/beer/def#
      brewery_id: https://github.com/dotwebstack/beer/id/brewery/
```

`dotwebstack.Rdf4j` is reserved by the Rdf4j backend as its properties location. The `shape.graph` property is used to 
point to the graph that is used within the `shapes.trig` file. The `shape.prefix` property is used to point to the prefix
of the shapesfile. This prefix is ignored when mapping GraphQL types from the `schema` to the node shapes in the
`shapes` file. Within the `prefixes` property you can define a list of general prefixes that is added to the 
[QueryEnvironment](https://github.com/dotwebstack/dotwebstack-framework/blob/v0.2/backend/Rdf4j/src/main/java/org/dotwebstack/framework/backend/Rdf4j/query/QueryEnvironment.java). 

> Besides the configuration in the `application.yml` you need to add a `schema` file that defines the GraphQL schema. 

Reviewer: 'schema.graphqls' ipv 'schema'

@prefix sh: <http://www.w3.org/ns/shacl#> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .
@prefix beer_def: <https://github.com/dotwebstack/beer/def#> .
@prefix beer_sh: <https://github.com/dotwebstack/beer/shapes#> .

<https://github.com/dotwebstack/beer/shapes> {
  beer_sh:Brewery a sh:NodeShape ;
    sh:targetClass beer_def:Brewery ;
    sh:property
      beer_sh:Brewery_identifier 
  .
  
  beer_sh:Brewery_identifier a sh:PropertyShape ;
    sh:name "identifier" ;
    sh:path beer_def:identifier ;
    sh:minCount 1 ;
    sh:maxCount 1 ;
    sh:nodeKind sh:Literal ;
    sh:datatype xsd:string
  .
}
```

First of all, notice that the named graph in the shapes file `https://github.com/dotwebstack/beer/shapes` is the one 
that is referred to by the `shape.graph` property in the snippet from `application.yml`. Also notice that the uri from 
the `beer_sh:` prefix is the one that is referred to by the `shape.prefix` property in the `application.yml`. A GraphQL 
type from `schema` is matched on a `node shape` in the `shapes` file after stripping the prefix from the name 
of the node shape. This means that the `Brewery` from the `schema` file is mapped on the `beer_sh:Brewery` 

> node shape. Fields from `schema` file are mapped on property shapes from the `shapes` file.

Reviewer: are maped on property shapes from the 'shapes' file based on the 'sh:name' property

}
```

## 1.3 Sparql directive

> The Rdf4j backend introduces the `@sparql` directive, this directive is added to a query in the `schema` file. It allows 

Reviewer: ```suggestion
The Rdf4j backend introduces the `@sparql` directive which can be added to a query in the `schema` file.
```

you to specify and assign paging and sort parameters and the specific query to a repository.

## 1.4 Paging parameters

Within the `@sparql` directive it is possible to declare a `limit` and an `offset` parameter. Arguments for both of 
these parameters are treated as [jexl](http://commons.apache.org/proper/commons-jexl/) expressions. The results of 

> both expressions need to be an integer and it is validated that `limit > 1` and `offset > 0`. The resulting values

Reviewer: ```suggestion
Arguments for both of these parameters are treated as [jexl](http://commons.apache.org/proper/commons-jexl/) expressions and should evaluate to an integer.  The resulting limit should be greater than 1, and the resulting offset should be greater than 0.
```

of these parameters are used as paging parameters in the resulting sparql query.  

## 1.5 Sorting

Within the `@sparql` directive it is possible to declare an `orderBy` parameter. The argument of this parameter needs to 
be one or any array of objects containing a `field` and an `order` parameter. The `field` parameter needs to be
a string and the `order` parameter is an `enum` and needs to be either `ASC` or `DESC`. The resulting values of these 

> parameters are in the order by clause in the resulting sparql query.  

Reviewer: ```suggestion
parameters are used in the order by clause in the resulting sparql query.  
```
